ministryofjustice/laa-apply-for-criminal-legal-aid
# CRIMAP-115 Scope the steps journey to the application uuid

## Description of change
The task list has surfaced a pesky issue with the fact we were assuming always an application is in the session and the journey affects just that application.

However with the task list, the provider can enter in whatever application they want, and then initiate the journey for that application, meaning we have to assign that application to the session so it becomes “active“, so to speak.

Instead of that approach which can be fragile and give problems if the provider click through different applications (as it will be re-assigning the session with each one), in this PR I've refactored some code to scope all steps journey to the specific application record, so all routes identify uniquely the application record they modify.

Example of scoped routes are:

```
/steps/:id/client/has_partner
/steps/:id/client/details
/steps/:id/address/lookup/:address_id
/steps/:id/address/details/:address_id
```

Where `:id` is the UUID of the `crime_application` record, and when we have nested resources like with addresses, any other IDs are named (like `:address_id`) to avoid clashing.

## Link to relevant ticket
https://dsdmoj.atlassian.net/browse/CRIMAP-115

## Notes for reviewer
As this involve logical changes to the routes, please do `CrimeApplication.destroy_all` to start fresh otherwise existing applications might have old values in their `navigation_stack`, making the back link fail.

## How to manually test the feature
As above, start fresh. The developer tools has changed to link directly to the dashboard. Start a new application from the dashboard, it will generate a new record, you can also check in the footer said UUID. From the task list enter the client details journey and check all routes are scoped to this record UUID. You can at any point "save and come back later" which will take you to the dashboard, and start a new application, etc.

NOTE: despite some heavy manual testing (in addition to the automated tests) there might be some place I've missed that fail, but hopefully is easy to fix.
